build.gradle

```diff
@@ -14,12 +14,16 @@ repositories {
 }
 
 dependencies {
-    compile 'com.github.shyiko:ktlint:0.30.0'
-    compile 'me.cassiano:ktlint-html-reporter:0.2.0'
-    compileOnly 'org.jetbrains.kotlin:kotlin-gradle-plugin'
-    compileOnly 'com.android.tools.build:gradle:3.0.1'
-    testCompile 'junit:junit:4.12'
-    testRuntime 'org.jetbrains.kotlin:kotlin-gradle-plugin'
+    implementation 'com.github.shyiko:ktlint:0.30.0'
+    implementation 'me.cassiano:ktlint-html-reporter:0.2.0'
+    implementation 'org.jetbrains.kotlin:kotlin-gradle-plugin'
```

Comment:
I couldn't get it to work either with a lesser scope. Even putting it both in `compileOnly` and `testImplementation`.

My original reason for using `compileOnly` for all the gradle dependencies was I was worried about it leaking into the buildscript dependencies and changing the version of gradle plugins the end-user was trying to use.

Even with `implementation` it does seem to introduce a change into the results of `gradle buildEnvironment` for a project using this plugin:
```
classpath
+--- org.jmailen.gradle:kotlinter-gradle:1.21.0
|    +--- com.github.shyiko:ktlint:0.30.0
|    |    +--- org.jetbrains.kotlin:kotlin-stdlib:1.3.20 -> 1.3.21
vvv newly introduced dependency
...
|    \--- org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.20 -> 1.3.21
|         +--- org.jetbrains.kotlin:kotlin-gradle-plugin-api:1.3.21
|         |    +--- org.jetbrains.kotlin:kotlin-stdlib:1.3.21 (*)
...
^^^ newly introduced dependency
\--- org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin:1.3.21
     \--- org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.21 (*)
```

What do you think? It seems like it could bump the user's kotlin gradle plugin up a version unintentionally in certain circumstances?

Reply:
I'll experiment a bit and see what our options are. I agree it's worth the time to get a better understanding.

Reply:
I admit I've been unable to figure out why it seems to need that dependency in the `implementation` configuration. I was looking at [your competitor](https://github.com/JLLeitschuh/ktlint-gradle/blob/master/plugin/build.gradle.kts) ( ;-) ), and he is using `compileOnly` and `testImplementation`.

Reply:
I believe ktlint-gradle executes the ktlint tool as a javaexec for each task whereas kotlinter-gradle runs ktlint's underlying engine inside the gradle runtime. This might explain the dependency differences and how they might affect TestKit.

Reply:
I can describe what I do to test. I know it's manual and tedious, but it's otherwise hard to validate the plugin actually works on different project types given it's relationship both with the kotlin compiler and the kotlin jvm and android plugins it extends.

Change the kotlinter-gradle version to be my branch name.
`gradle publishToMavenLocal`

I have a couple of projects that use kotlinter-gradle, so I change the build.gradle to have:
```groovy
buildscript {
    repositories {
        mavenLocal()
        jcenter()
    }

    dependencies {
        classpath 'org.jmailen.gradle:kotlinter-gradle:my-branch-name'
    }
}
```

On the project that uses it, I run through any fresh and incremental linting and formatting tests I want to do. It's not as terrible as it sounds once you get it set up the first time.

Reply:
I was able to resolve the issue (with inspiration from [here](https://github.com/JLLeitschuh/ktlint-gradle/blob/master/plugin/build.gradle.kts#L18-L20)).